File: snippets/python-2025-08-01/create-device.py
# @title: Create Device
# @description: Create new devices programmatically using dashboard inputs
# @tags: device, create, dashboard, automation

# /// script
# dependencies = [
#   "tagoio-sdk"
# ]
# ///

"""
Analysis Example
Creating devices using dashboard

Using an Input Widget in the dashboard, you will be able to create devices in your account.
You can get the dashboard template to use here: https://admin.tago.io/template/6143555a314cef001871ec78
Use a dummy HTTPs device with the dashboard.

Environment Variables
In order to use this analysis, you must setup the Environment Variable table.
  account_token: Your account token. Check bellow how to get this.

Steps to generate an account_token:
1 - Enter the following link: https://admin.tago.io/account/
2 - Select your Profile.
3 - Enter Tokens tab.
4 - Generate a new Token with Expires Never.
5 - Press the Copy Button and place at the Environment Variables tab of this analysis.
"""

from tagoio_sdk import Analysis, Device, Account
from tagoio_sdk.modules.Utils.getTokenByName import getTokenByName
from tagoio_sdk.modules.Account.Device_Type import DeviceCreateInfo


def add_configuration_parameter_to_device(account: Account, device_id: str) -> None:
    account.devices.paramSet(
        deviceID=device_id, configObj={ "key": "param_key", "value": "10", "sent": False }
    )


def send_feedback_to_dashboard(account: Account, device_id: str) -> None:
    dashboard_token = getTokenByName(account=account, deviceID=device_id)
    device = Device(params={"token":dashboard_token})

    # To add any data to the device that was just created:
    # device.sendData({ "variable": "temperature", value: 17 })

    device.sendData(
        data={"variable": "validation", "value": "Device successfully created!", "metadata": {"type": "success" } }
    )


def parse_new_device(scope: list[dict]) -> DeviceCreateInfo:
    # Get the variables sent by the widget/dashboard.
    device_network = [obj for obj in scope if obj["variable"]  == "device_network"]
    device_connector = [obj for obj in scope if obj["variable"]  == "device_connector"]
    device_name = [obj for obj in scope if obj["variable"]  == "device_name"]
    device_eui = [obj for obj in scope if obj["variable"]  == "device_eui"]

    if not device_network or not device_network[0]["value"]:
        raise TypeError('Missing "device_network" in the data scope.')
    elif not device_connector or not device_connector[0]["value"]:
        raise TypeError('Missing "device_connector" in the data scope.')
    elif not device_eui or not device_eui[0]["value"]:
        raise TypeError('Missing "device_eui" in the data scope.')

    return {
        "name": device_name[0]["value"],
        "serie_number": device_eui[0]["value"],
        "tags": [
            # You can add custom tags here.
            { "key": "type", "value": "sensor" },
            { "key": "device_eui", "value": device_eui[0]["value"] },
        ],
        "connector": device_connector[0]["value"],
        "network": device_network[0]["value"],
        "active": True,
        "type": "immutable",
        "chunk_period": "month", # consider change
        "chunk_retention": 1, # consider change
    }


def start_analysis(context: list[dict], scope: list[dict]) -> None:
    if not scope:
        return print("The analysis must be triggered by a widget.")

    # reads the value of account_token from the environment variable
    account_token = list(filter(lambda account_token: account_token["key"] == "account_token", context.environment))
    account_token = account_token[0]["value"]

    if not account_token:
        return print("Missing account_token Environment Variable.")

    account = Account(params={"token": account_token})

    new_device = parse_new_device(scope=scope)

    result = account.devices.create(deviceObj=new_device)
    print(result)

    add_configuration_parameter_to_device(account=account, device_id=result["device_id"])

    send_feedback_to_dashboard(account=account, device_id=scope[0]["device"])


# The analysis token in only necessary to run the analysis outside TagoIO
Analysis(params={"token": "MY-ANALYSIS-TOKEN-HERE"}).init(start_analysis)

File: snippets/deno-2025-08-01/create-device.ts
// @title: Create Device from Dashboard
// @description: Create new devices dynamically using dashboard interface
// @tags: device, create, dashboard, dynamic, management

/*
 * Example: Creating Devices via Dashboard
 * This example demonstrates how to create devices in your account using an Input Widget on the dashboard.
 *
 * Dashboard Template:
 * You can access the dashboard template needed for this operation here: https://admin.tago.io/template/6143555a314cef001871ec78
 * It's recommended to use a dummy HTTPS device alongside the dashboard for testing purposes.
 *
 * Usage Instructions:
 * For the analysis to function correctly, you must add a new policy to your account by following these steps:
 *  1. Navigate to https://admin.tago.io/am and click on the "Add Policy" button.
 *  2. In the Target selector, ensure the field is set to "ID", then select your Analysis from the list.
 *  3. Click on the "Click to add a new permission" option, choose "Device" as the type, and set the rule to "Access" with the scope as "Any".
 *  4. Finalize by clicking the save button located in the bottom right corner to apply your new Policy.
 */

import { Analysis, Resources } from "jsr:@tago-io/sdk";
import type { AnalysisConstructorParams, Data } from "jsr:@tago-io/sdk";

async function startAnalysis(context: AnalysisConstructorParams, scope: Data[]): Promise<void> {
  if (!scope[0]) {
    return console.log("The analysis must be triggered by a widget.");
  }

  console.log("Creating your device");

  // Get the variables sent by the widget/dashboard.
  const network_id = scope.find((x) => x.variable === "device_network");
  const connector_id = scope.find((x) => x.variable === "device_connector");
  const device_name = scope.find((x) => x.variable === "device_name");
  const device_eui = scope.find((x) => x.variable === "device_eui");

  if (!connector_id || !connector_id.value) {
    return console.log('Missing "device_connector" in the data scope.');
  } else if (!network_id || !network_id.value) {
    return console.log('Missing "device_network" in the data scope.');
  } else if (!device_eui || !device_eui.value) {
    return console.log('Missing "device_eui" in the data scope.');
  }

  const deviceID = scope[0]?.device;
  if (!deviceID) {
    return console.log('Device ID not found in the data scope');
  }

  const deviceCreateInfo: any = {
    name: device_name.value as string,
    // Serie number is the parameter for device eui, sigfox id, etc..
    serie_number: device_eui.value as string,
    tags: [
      // You can add custom tags here.
      { key: "type", value: "sensor" },
      { key: "device_eui", value: device_eui.value as string },
    ],
    connector: connector_id.value as string,
    network: network_id.value as string,
    active: true,
    type: "immutable",
    chunk_period: "month", //consider change
    chunk_retention: 1, //consider change
  };

  const result = await Resources.devices
    .create(deviceCreateInfo)
    .catch((error) => {
      // Send the validation to the device.
      // That way we create an error in the dashboard for feedback.
      Resources.devices.sendDeviceData(deviceID, {
        variable: "validation",
        value: `Error when creating the device ${error}`,
        metadata: { color: "red" },
      });
      throw error;
    });

  // To add Configuration Parameters to the device:
  await Resources.devices.paramSet(result.device_id, { key: "param_key", value: "10", sent: false });

  // Send feedback to the dashboard:
  await Resources.devices.sendDeviceData(deviceID, { variable: "validation", value: "Device succesfully created!", metadata: { type: "success" } });
  console.log(`Device succesfully created. ID: ${result.device_id}`);
}

Analysis.use(startAnalysis);